casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala

```diff
@@ -237,34 +237,41 @@ object LfsBlockRequester {
           _ <- st.update(_.done(block.blockHash))
         } yield ()
 
-      import cats.instances.list._
-
       /**
-        * Request stream is pulling new block hashes ready for broadcast requests.
+        * Reads current state for next blocks and send the requests.
+        *
+        * @param resend should re-request already requested blocks
         */
-      val requestStream = for {
-        // Request queue is a trigger when to check the state
-        resend <- requestQueue.dequeueChunk(maxSize = 1)
+      def requestNext(resend: Boolean) =
+        for {
+          // Check if stream is finished (no more requests)
+          isEnd <- st.get.map(_.isFinished)
 
-        // Check if stream is finished (no more requests)
-        isEnd <- Stream.eval(st.get.map(_.isFinished))
+          // Take next set of items to request (w/o duplicates)
+          hashes <- st.modify(_.getNext(resend))
 
-        // Take next set of items to request (w/o duplicates)
-        hashes <- Stream.eval(st.modify(_.getNext(resend)))
+          // Check existing blocks
+          existingHashes <- hashes.toList.filterA(containsBlock)
 
-        // Check existing blocks
-        existingHashes <- Stream.eval(hashes.toList.filterA(containsBlock))
+          // Enqueue hashes of exiting blocks
+          _ <- responseHashQueue
+                .enqueue(Stream.emits(existingHashes))
+                .compile
+                .drain
+                .whenA(existingHashes.nonEmpty)
```

Comment:
Here was the bug, when flatMap-ed, if Stream has more then one element it will execute rest of the comprehension for each produced element.